rclc/src/functions.c

```diff
@@ -0,0 +1,255 @@
+#include <rclc/rclc.h>
+
+#include <rcl/rcl.h>
+#include <rcl/error_handling.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+/*/
+#define ALLOCATE(s) rcl_get_default_allocator().allocate(s, rcl_get_default_allocator().state)
+#define DEALLOCATE(ptr) rcl_get_default_allocator().deallocate(ptr, rcl_get_default_allocator().state)
+#define REALLOCATE(ptr, s) rcl_get_default_allocator().reallocate(ptr, s, rcl_get_default_allocator().state)
+/*/
+#define ALLOCATE(s) malloc(s)
+#define DEALLOCATE(ptr) free(ptr)
+#define REALLOCATE(ptr, s) realloc(ptr, s)
+//*/
+
+struct rclc_subscription_t
+{
+  rcl_subscription_t rcl_subscription;
+  rclc_callback_t user_callback;
+
+  rclc_node_t * node;
+};
+
+struct rclc_publisher_t
+{
+  rcl_publisher_t rcl_publisher;
+
+  rclc_node_t * node;
+};
+
+struct rclc_node_t
+{
+  rcl_node_t rcl_node;
+
+  rclc_subscription_t ** subs;
+  size_t subs_s;
+};
+
+rclc_ret_t rclc_init(int argc, char ** argv)
+{
+  rcl_ret_t ret = rcl_init(argc, argv, rcl_get_default_allocator());
+  if(ret != RCL_RET_OK) {
+    fprintf(stderr, "[rclc_init] error in rcl_init: %s\n", rcl_get_error_string_safe());
+    rcl_reset_error();
+  }
+  return ret;
+}
+
+bool rclc_ok()
+{
+  return true;
+}
+
+void rclc_sleep_ms(size_t milliseconds)
```

Comment:
Yes it is used in the [publisher example](https://github.com/astralien3000/examples/blob/master/rclc/minimal_publisher/not_composable.c). But it is tricky. Is changed the implementation compared to beta1, it was supposed to do almost the same things than `rclc_spin_node`, but now it does not. I don't know what rcl_wait do exactly so that I kept it, but I will try to remove it.

I think it is the most "outdated" part of the API, anyway the function should change.

Reply:
Yeah I think we would need to update the talker example:
https://github.com/ros2/tutorials/blob/master/rclc_tutorials/src/talker.c
to use a timer like the [Python](https://github.com/ros2/demos/blob/master/demo_nodes_py/demo_nodes_py/topics/talker.py) and [c++](https://github.com/ros2/demos/blob/master/demo_nodes_cpp/src/topics/talker.cpp) ones. The difference being that python uses an rcl timer (what I think rclc should use) while rclcpp uses a chrono timer

Reply:
Since I have not implemented the timer part, I will go for replacing rclc_sleep_ms by rclc_spin_node_once, enabling user to specify a timeout.